src/feature/hs/hs_service.c

```diff
@@ -4103,6 +4103,41 @@ hs_service_load_all_keys(void)
   return -1;
 }
 
+/** Log the status of introduction points for all version 3 onion services
+ * at log severity <b>severity</b>.
+ */
+void
+hs_service_dump_stats(int severity)
+{
+  origin_circuit_t *circ;
+
+  for (hs_service_t **iter = HT_START(hs_service_ht, hs_service_map);
```

Comment:
Please check out `FOR_EACH_SERVICE_BEGIN`. For example see how `update_all_descriptors_intro_points()` does it. You've already used `FOR_EACH_DESCRIPTOR_BEGIN` so it should be similar.